# Notebook 3 – Measures of Dispersion

# 1. Range

## Step 1 – Understand the Concept

Range is the difference between the maximum value and the minimum value in a dataset.

It gives us a simple idea of how widely the values are spread.

A larger range means the values have a wider spread, while a smaller range means the values are closer together.

### Formula

Range = Maximum Value - Minimum Value

## Step 2 – Demonstrate the Concept

### Numerical Example

Consider the following daily sales:

100, 120, 150, 180, 200

Maximum value = 200
Minimum value = 100

Range = 200 - 100

Range = 100

Therefore, the sales data has a range of 100 units.

### Real-World Example

Suppose a delivery company records delivery times:

30, 35, 40, 45, 60 minutes.

The range is:

60 - 30 = 30 minutes.

This tells the company that the difference between the fastest and slowest delivery is 30 minutes.

### Business Example

A business can use range to quickly understand the difference between its lowest and highest daily sales.

### AI/ML Example

Range can be used during Exploratory Data Analysis (EDA) to understand the spread of a numerical feature.

### How does this demonstrate the concept?

Range compares only the smallest and largest values and shows the total spread between them.

## Step 3 – Implement the Concept

In [1]:
import numpy as np

sales = np.array([100, 120, 150, 180, 200])

data_range = np.max(sales) - np.min(sales)

print("Maximum:", np.max(sales))
print("Minimum:", np.min(sales))
print("Range:", data_range)

Maximum: 200
Minimum: 100
Range: 100


## Interpretation

### Output

Maximum: 200
Minimum: 100
Range: 100

### What does the program do?

The program finds the maximum and minimum sales and calculates their difference.

### What statistical concept does it demonstrate?

It demonstrates the range of the dataset.

### How do we interpret the output?

The sales values are spread across a range of 100 units.

### Where is it used in AI/ML?

Range is useful during EDA to quickly understand the spread of numerical features.

# 2. Variance

## Step 1 – Understand the Concept

Variance measures how far the values in a dataset are spread from the mean.

It calculates the average of the squared differences between each value and the mean.

A small variance means the values are close to the mean.

A large variance means the values are more spread out.

### Population Variance Formula

Variance = Σ(X - μ)² / N

Where:

X = Individual value

μ = Population mean

N = Number of observations

The differences are squared so that negative and positive differences do not cancel each other.

## Step 2 – Demonstrate the Concept

### Numerical Example

Consider:

10, 20, 30

Mean:

(10 + 20 + 30) / 3 = 20

Differences from mean:

10 - 20 = -10
20 - 20 = 0
30 - 20 = 10

Squared differences:

100, 0, 100

Variance:

(100 + 0 + 100) / 3

= 66.67

### Real-World Example

Suppose two shops have the same average daily sales.

Shop A:
100, 100, 100

Shop B:
50, 100, 150

Both have an average of 100, but Shop B has much more variation.

Variance helps identify this difference.

### Business Example

A company can use variance to understand how much daily sales fluctuate around the average.

### AI/ML Example

Variance is useful during feature analysis because features with very different levels of variation may require preprocessing such as scaling.

### How does this demonstrate the concept?

Variance measures the average squared distance of the observations from the mean.


## Step 3 – Implement the Concept

In [2]:
import numpy as np

sales = np.array([10, 20, 30])

variance = np.var(sales)

print("Mean:", np.mean(sales))
print("Variance:", variance)

Mean: 20.0
Variance: 66.66666666666667


## Interpretation

### Output

Mean: 20.0
Variance: 66.66666666666667

### What does the program do?

The program calculates the mean and population variance of the sales data.

### What statistical concept does it demonstrate?

It demonstrates variance as a measure of data spread.

### How do we interpret the output?

The variance is approximately 66.67.

This indicates how much the values vary around the mean.

### Where is it used in AI/ML?

Variance is useful in feature analysis, statistical modeling, and understanding data variability.

# 3. Standard Deviation

## Step 1 – Understand the Concept

Standard deviation measures how much the values in a dataset typically differ from the mean.

It is the square root of variance.

Unlike variance, standard deviation is expressed in the same unit as the original data, which makes it easier to interpret.

### Formula

Standard Deviation = √Variance

For population data:

σ = √[Σ(X - μ)² / N]

## Step 2 – Demonstrate the Concept

### Numerical Example

Data:

10, 20, 30

Mean = 20

Population variance = 66.67

Standard deviation:

√66.67 ≈ 8.16

Therefore, the standard deviation is approximately 8.16.

### Real-World Example

Suppose two machines produce bottles with the same average weight.

Machine A produces:

500, 501, 499, 500

Machine B produces:

450, 550, 500, 500

Machine B has a larger spread.

Standard deviation helps identify which machine has more variation in production.

### Business Example

A company can use standard deviation to understand fluctuations in sales, delivery times, or customer spending.

### AI/ML Example

Standard deviation is commonly used when standardizing numerical features.

For example, StandardScaler uses the mean and standard deviation to transform features.

### How does this demonstrate the concept?

Standard deviation converts the spread measured by variance back into the original unit of the data.


## Step 3 – Implement the Concept

In [3]:
import numpy as np

sales = np.array([10, 20, 30])

standard_deviation = np.std(sales)

print("Mean:", np.mean(sales))
print("Standard Deviation:", standard_deviation)

Mean: 20.0
Standard Deviation: 8.16496580927726


## Interpretation

### Output

Mean: 20.0
Standard Deviation: 8.16496580927726

### What does the program do?

The program calculates the standard deviation of the sales values.

### What statistical concept does it demonstrate?

It demonstrates the spread of data around the mean.

### How do we interpret the output?

The standard deviation is approximately 8.16 units.

A larger standard deviation would indicate greater variation in the data.

### Where is it used in AI/ML?

Standard deviation is used in feature scaling, normalization, statistical analysis, and model preprocessing.

# 4. Quartiles

## Step 1 – Understand the Concept

Quartiles divide an ordered dataset into four parts.

There are three main quartiles:

- Q1 – First Quartile: 25% of the data is below this value.
- Q2 – Second Quartile: 50% of the data is below this value. It is the median.
- Q3 – Third Quartile: 75% of the data is below this value.

Quartiles help us understand the distribution and position of values in a dataset.

## Step 2 – Demonstrate the Concept

### Numerical Example

Consider the ordered data:

10, 20, 30, 40, 50, 60, 70, 80

Q1 represents approximately the 25th percentile.

Q2 represents the 50th percentile, which is the median.

Q3 represents approximately the 75th percentile.

### Real-World Example

Suppose a company studies employee salaries.

Q1 tells us the salary level below which approximately 25% of employees fall.

Q2 represents the middle salary.

Q3 tells us the salary level below which approximately 75% of employees fall.

### Business Example

A company can use quartiles to divide customers according to spending levels.

For example, Q1 can help identify lower-spending customers, while Q3 can help identify higher-spending customers.

### AI/ML Example

Quartiles are useful during EDA to understand feature distributions and identify possible outliers.

### How does this demonstrate the concept?

Quartiles divide the ordered data into four sections and help us understand where observations are located within the distribution.


## Step 3 – Implement the Concept

In [4]:
import numpy as np

data = np.array([10, 20, 30, 40, 50, 60, 70, 80])

q1 = np.percentile(data, 25)
q2 = np.percentile(data, 50)
q3 = np.percentile(data, 75)

print("Q1:", q1)
print("Q2:", q2)
print("Q3:", q3)

Q1: 27.5
Q2: 45.0
Q3: 62.5


## Interpretation

### What does the program do?

The program calculates the first, second, and third quartiles using NumPy.

### What statistical concept does it demonstrate?

It demonstrates how quartiles divide data based on position.

### How do we interpret the output?

Q1 represents the 25th percentile.

Q2 represents the 50th percentile or median.

Q3 represents the 75th percentile.

### Where is it used in AI/ML?

Quartiles are used during EDA to understand feature distributions and identify potential outliers.

# 5. Percentiles

## Step 1 – Understand the Concept

A percentile tells us the value below which a certain percentage of observations fall.

For example:

- 25th percentile → 25% of values are below it.
- 50th percentile → 50% of values are below it.
- 90th percentile → 90% of values are below it.

Percentiles help us understand the relative position of a value within a dataset.

## Step 2 – Demonstrate the Concept

### Numerical Example

Suppose the test scores are:

50, 60, 65, 70, 75, 80, 85, 90, 95, 100

If a student's score is around the 90th percentile, it means the student's score is higher than approximately 90% of the observations in the dataset.

### Real-World Example

Suppose a student is at the 90th percentile in an entrance exam.

This means the student's performance is at or above the level of approximately 90% of the students in the reference dataset, depending on the percentile calculation method.

### Business Example

An e-commerce company may use the 95th percentile of delivery time to understand a high-end typical delivery duration and monitor unusually slow deliveries.

### AI/ML Example

Percentiles can be used to detect extreme values and understand the distribution of numerical features.

### How does this demonstrate the concept?

A percentile describes the relative position of a value compared with the rest of the dataset.


## Step 3 – Implement the Concept

In [5]:
import numpy as np

scores = np.array([50, 60, 65, 70, 75, 80, 85, 90, 95, 100])

p25 = np.percentile(scores, 25)
p50 = np.percentile(scores, 50)
p90 = np.percentile(scores, 90)

print("25th Percentile:", p25)
print("50th Percentile:", p50)
print("90th Percentile:", p90)

25th Percentile: 66.25
50th Percentile: 77.5
90th Percentile: 95.5


##  Interpretation

### What does the program do?

The program calculates the 25th, 50th, and 90th percentiles of the scores.

### What statistical concept does it demonstrate?

It demonstrates how percentiles describe the position of observations within a dataset.

### How do we interpret the output?

The 90th percentile represents the score below which approximately 90% of the dataset lies.

### Where is it used in AI/ML?

Percentiles are useful for outlier detection, feature analysis, and understanding skewed distributions.

# 6. Interquartile Range (IQR)

## Step 1 – Understand the Concept

Interquartile Range, or IQR, measures the spread of the middle 50% of the data.

It is calculated as the difference between the third quartile and the first quartile.

### Formula

IQR = Q3 - Q1

Where:

Q1 = 25th percentile

Q3 = 75th percentile

IQR is useful because it is not strongly affected by extreme values.

## Step 2 – Demonstrate the Concept

### Numerical Example

Consider:

10, 20, 30, 40, 50, 60, 70, 80

Suppose:

Q1 = 25
Q3 = 65

Then:

IQR = Q3 - Q1

IQR = 65 - 25

IQR = 40

Therefore, the middle 50% of the data has a spread of 40 units.

### Real-World Example

Suppose a company studies employee salaries.

IQR tells the company how widely the middle 50% of employee salaries are distributed.

### Business Example

An e-commerce company can use IQR to understand the spread of the middle 50% of customer spending values without being heavily influenced by a few extremely high-spending customers.

### AI/ML Example

IQR is commonly used for detecting outliers.

A common rule is:

Lower Bound = Q1 - 1.5 × IQR

Upper Bound = Q3 + 1.5 × IQR

Values outside these boundaries can be considered potential outliers.

### How does this demonstrate the concept?

IQR focuses only on the middle 50% of observations, making it useful for measuring spread without being strongly influenced by extreme values.


## Step 3 – Implement the Concept

In [6]:
import numpy as np

data = np.array([10, 20, 30, 40, 50, 60, 70, 80, 200])

q1 = np.percentile(data, 25)
q3 = np.percentile(data, 75)

iqr = q3 - q1

print("Q1:", q1)
print("Q3:", q3)
print("IQR:", iqr)

Q1: 30.0
Q3: 70.0
IQR: 40.0


## Interpretation

### What does the program do?

The program calculates Q1 and Q3 and then subtracts Q1 from Q3 to obtain the IQR.

### What statistical concept does it demonstrate?

It demonstrates the spread of the middle 50% of the dataset.

### How do we interpret the output?

The IQR shows how widely the middle half of the observations are spread.

The extreme value 200 has less influence on IQR than it would have on the range.

### Where is it used in AI/ML?

IQR is commonly used during EDA for identifying potential outliers and understanding feature distributions.

# 7. Coefficient of Variation

## Step 1 – Understand the Concept

Coefficient of Variation (CV) measures the amount of variation relative to the mean.

It is expressed as a percentage.

CV is especially useful when comparing the variability of datasets that have different means or different scales.

### Formula

CV = (Standard Deviation / Mean) × 100

A lower CV means less variation relative to the mean.

A higher CV means more variation relative to the mean.

## Step 2 – Demonstrate the Concept

### Numerical Example

Dataset A:

Mean = 100
Standard Deviation = 10

CV = (10 / 100) × 100

CV = 10%

Dataset B:

Mean = 200
Standard Deviation = 30

CV = (30 / 200) × 100

CV = 15%

Dataset B has greater relative variation.

### Real-World Example

Suppose two machines produce components.

Machine A:
Average weight = 100 grams
Standard deviation = 5 grams

Machine B:
Average weight = 200 grams
Standard deviation = 8 grams

The standard deviations alone cannot directly tell us which machine has greater relative variation because their average weights are different.

Coefficient of Variation allows us to compare variation relative to the mean.

### Business Example

A company can compare the consistency of sales, production output, or delivery times across products with different average values.

### AI/ML Example

CV can help compare variability between numerical features with different scales.

### How does this demonstrate the concept?

CV converts standard deviation into a percentage of the mean, making relative variability easier to compare across datasets.


## Step 3 – Implement the Concept

In [7]:
import numpy as np

data_a = np.array([90, 95, 100, 105, 110])
data_b = np.array([170, 185, 200, 215, 230])

mean_a = np.mean(data_a)
std_a = np.std(data_a)

mean_b = np.mean(data_b)
std_b = np.std(data_b)

cv_a = (std_a / mean_a) * 100
cv_b = (std_b / mean_b) * 100

print("Dataset A")
print("Mean:", mean_a)
print("Standard Deviation:", std_a)
print("Coefficient of Variation:", cv_a, "%")

print("\nDataset B")
print("Mean:", mean_b)
print("Standard Deviation:", std_b)
print("Coefficient of Variation:", cv_b, "%")

Dataset A
Mean: 100.0
Standard Deviation: 7.0710678118654755
Coefficient of Variation: 7.0710678118654755 %

Dataset B
Mean: 200.0
Standard Deviation: 21.213203435596427
Coefficient of Variation: 10.606601717798213 %


## Interpretation

### What does the program do?

The program calculates the mean, standard deviation, and coefficient of variation for two datasets.

### What statistical concept does it demonstrate?

It demonstrates relative variability using the coefficient of variation.

### How do we interpret the output?

The CV expresses the standard deviation as a percentage of the mean.

A lower CV indicates lower variation relative to the mean.

A higher CV indicates higher variation relative to the mean.

### Where is it used in AI/ML?

CV can help compare the relative variability of numerical features that have different scales.